docs/proposal/weight_policies_aws.md

```diff
@@ -0,0 +1,204 @@
+under its ownership
+
+## Types of Ownship and Scope of Feature
+
+For now this will focus on multiple records per the same owner, referred to as "cluster" ownership. In the future
+the plan will be to support multiple records for multiple owners, referred to as "global" ownership.
+
+## Code Changes
+
+### Extend the Endpoint Type to Include Policies
+
+The endpoint.Endpoint type will include a Policy like so:
+
+```go
+type Endpoint struct {
+  // The hostname of the DNS record
+  DNSName string
+  // The target the DNS record points to
+  Target string
+  // RecordType type of record, e.g. CNAME, A, TXT etc
+  RecordType string
+  // Policy stores policies for the endpoint
+  Policy EndpointPolicy
+  // Labels stores labels defined for the Endpoint
+  Labels map[string]string
+}
+```
+
+The EndpointPolicy may include a number of policies depending on the source. In the case the source has the appropriate
+annotations, it may include a WeightPolicy like so:
+
+```go
+type EndpointPolicy struct {
+  // Weight is the weight policy for the endpoint
+  Weight WeightPolicy
+}
+```
+
+### Manage Weight Policies as an Interface
+
+We'll assume there will be other types other than AWS ELBs that will need weights as well. So WeightPolicy is an
+interface like:
+
+```go
+type Weight interface {
+  Amount() int64
+  ID() string
+}
+```
+
+The route53WeightPolicy type implments the WeightPolicy interface. It is specific to Route53 and can be created via the
+following:
+
+```go
+type route53WeightPolicy struct {
+  // scope is can be cluster or none
+  // TODO: Implement global weight scope in addition to cluster-only
+  scope string
+  // weight is the weight of the record
+  weight int64
+  // setidentifier for the weight policy
+  setIdentifier string
+}
+
+func NewRoute53WeightPolicy(weight int64, setIdentifier string) Weight {
+  return route53WeightPolicy{
+    scope: parseScope(setIdentifier),
+    weight: weight,
+    setIdentifier: setIdentifier,
+  }
+}
+```
+
+### How to Set Weight Policies from the Provider
+
+Policies can be added to an endpoint through methods on the Endpoint type. For example, inside the AWS provider:
+
+```go
+newEndpoint := endpoint.NewEndpoint(wildcardUnescape(aws.StringValue(r.Name)), aws.StringValue(rr.Value), aws.StringValue(r.Type))
+newEndpoint.AttachWeightPolicy(endpoint.policies.NewRoute53WeightPolicy(aws.Int64Value(r.Weight), aws.aws.StringValue(r.SetIdentifier)))
+endpoints = append(endpoints, newEndpoint)
+```
+
+This assumes `r.SetIdentifier` is something like `cluster/service-a` or `global/us-east-1`. These are set in the next
+section.
+
+### How to Set Weight Polcies from the Source
+
+Policies can be added to an endpoint from the Source in the same fashion as the provider, but using values specified via
+the following annotations:
+
+```go
+  // The annotation used for defining the desired weight for the record
+  weightAnnotationKey = "external-dns.alpha.kubernetes.io/weight"
```

Comment:
I think these annotations should be namespaced to their cloud provider, similiar to annotations for controlling AWS ELBs.

For example, `service.beta.kubernetes.io/aws-load-balancer-proxy-protocol` is the annotation for specifying proxy protocol on LoadBalancer services for AWS.

These annotations should probably be something like
```
external-dns.alpha.kubernetes.io/aws-weight
```

This ensures these values are only ever for AWS, because weighting may be done differently per cloud provider. Additionally, since these control routing policy, perhaps we should name them like this:

```
external-dns.alpha.kubernetes.io/aws-routing-policy-weight
external-dns.alpha.kubernetes.io/aws-routing-policy-weight-scope
```

This ensures you can add other routing policies in the future, such as failover routing policies, or latency based routing policies.

Reply:
The purpose of weight-scope is to provide info on how to manage the ownership (TXT records). Scope is not an attribute that AWS is aware of. So I'm wondering if we should keep this one as
```
external-dns.alpha.kubernetes.io/weight-scope
```

You make good points about AWS-specific settings. These exist in route53 under a resource record set as per https://docs.aws.amazon.com/Route53/latest/APIReference/API_ResourceRecordSet.html. So I'm wondering if we'd want to do this:

```
external-dns.alpha.kubernetes.io/aws-route53-resource-record-set-weight
external-dns.alpha.kubernetes.io/aws-route53-resource-record-set-set-identifier
```

That way in the future we could add for example:
```
external-dns.alpha.kubernetes.io/aws-route53-resource-record-set-failover
external-dns.alpha.kubernetes.io/aws-route53-resource-record-set-geo-location
```
What do you think?

Reply:
Ah, I thought the scope portion was related to the identifier, my mistake. 

Yeah, that would probably work fine, I think we could potentially abbreviate `resource-record-set` but, it's not a huge deal either way, so long as we're namespacing annotations adequately.